File: packages/go/libsql/pk_test.go
package libsql_test

import (
	"context"
	"database/sql"
	"path/filepath"
	"strings"
	"testing"

	"github.com/b4moss/crudian/go/crudian"
	libsqlcrud "github.com/b4moss/crudian/go/libsql"
	_ "github.com/tursodatabase/libsql-client-go/libsql"
	_ "modernc.org/sqlite" // required by libsql-client-go for file:// URLs
)

func openItemIDDB(t *testing.T) *sql.DB {
	t.Helper()
	path, err := filepath.Abs(filepath.Join(t.TempDir(), "pk.db"))
	if err != nil {
		t.Fatalf("abs: %v", err)
	}
	db, err := sql.Open("libsql", "file://"+filepath.ToSlash(path))
	if err != nil {
		t.Fatalf("open: %v", err)
	}
	t.Cleanup(func() { _ = db.Close() })
	if _, err := db.Exec(`CREATE TABLE items (
		item_id INTEGER PRIMARY KEY AUTOINCREMENT,
		name TEXT NOT NULL,
		score INTEGER NOT NULL,
		note TEXT
	)`); err != nil {
		t.Fatalf("migrate: %v", err)
	}
	return db
}

func TestConfigurablePKSmoke(t *testing.T) {
	ctx := context.Background()
	crud, err := libsqlcrud.CreateCrud(openItemIDDB(t), crudian.Options{PK: "item_id"})
	if err != nil {
		t.Fatalf("CreateCrud: %v", err)
	}
	created, err := crud.Create(ctx, "items", crudian.Row{"name": "a", "score": 1})
	if err != nil || toInt64(created["item_id"]) == 0 {
		t.Fatalf("create: %+v err=%v", created, err)
	}
	page, err := crud.Search(ctx, "items", crudian.SearchQuery{Paging: "offset", Limit: 10})
	if err != nil || len(page.Items) != 1 {
		t.Fatalf("search: %+v err=%v", page, err)
	}
	_, err = crud.Upsert(ctx, "items", crudian.Row{"name": "missing-pk"})
	if err == nil || !strings.Contains(err.Error(), "item_id") {
		t.Fatalf("expected upsert pk error, got %v", err)
	}
}

File: packages/js/src/typeorm/pk.test.ts
import assert from "node:assert/strict"
import { mkdtempSync, rmSync } from "node:fs"
import { tmpdir } from "node:os"
import { join } from "node:path"
import { after, before, describe, test } from "node:test"
import "reflect-metadata"
import { DataSource } from "typeorm"
import { CrudianError, where } from "../index.js"
import { createCrud } from "./index.js"

const ID_SCHEMA = `
  CREATE TABLE items (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    name TEXT NOT NULL,
    score INTEGER,
    note TEXT
  );
`

const ITEM_ID_SCHEMA = `
  CREATE TABLE items_pk (
    item_id INTEGER PRIMARY KEY AUTOINCREMENT,
    name TEXT NOT NULL,
    score INTEGER,
    note TEXT
  );
`

let tempDir = ""
let dataSource: DataSource

before(async () => {
  tempDir = mkdtempSync(join(tmpdir(), "crudian-typeorm-pk-"))
  const dbPath = join(tempDir, "test.db")
  dataSource = new DataSource({
    type: "better-sqlite3",
    database: dbPath,
  })
  await dataSource.initialize()
  await dataSource.query(ID_SCHEMA)
  await dataSource.query(ITEM_ID_SCHEMA)
})

after(async () => {
  if (dataSource?.isInitialized) await dataSource.destroy()
  if (tempDir) rmSync(tempDir, { recursive: true, force: true })
})

describe("typeorm.createCrudPk", () => {
  test("正常系: カスタム PK で create / search paging / upsert / duplicate", async () => {
    await dataSource.query("DELETE FROM items_pk")
    const crud = createCrud(dataSource, { pk: "item_id" })
    const created = await crud.create("items_pk", { name: "a", score: 1 })
    assert.equal(typeof created.item_id, "number")

    await crud.create("items_pk", { name: "b", score: 1 })
    await crud.create("items_pk", { name: "c", score: 1 })
    const page = await crud.search("items_pk", { paging: "cursor", limit: 2 })
    assert.equal(page.hasMore, true)
    assert.ok("nextCursor" in page)
    assert.equal(page.nextCursor, 2)

    const upserted = await crud.upsert("items_pk", {
      item_id: 10,
      name: "x",
      score: 9,
    })
    assert.equal(upserted.item_id, 10)

    const dup = await crud.duplicate("items_pk", {
      where: where().eq("item_id", created.item_id),
      overrides: { name: "dup" },
    })
    assert.ok(dup)
    assert.notEqual(dup!.item_id, created.item_id)
    assert.equal(dup!.name, "dup")
  })

  test("正常系: 省略時は id デフォルト", async () => {
    await dataSource.query("DELETE FROM items")
    const crud = createCrud(dataSource)
    const row = await crud.create("items", { name: "a", score: 1 })
    assert.equal(row.id, 1)
  })

  test("異常系: 空 pk・列不存在・upsert PK 欠落", async () => {
    assert.throws(() => createCrud(dataSource, { pk: "" }), CrudianError)

    const missing = createCrud(dataSource, { pk: "item_id" })
    await assert.rejects(
      () => missing.create("items", { name: "a" }),
      CrudianError,
    )

    const crud = createCrud(dataSource, { pk: "item_id" })
    await assert.rejects(
      () => crud.upsert("items_pk", { name: "a" }),
      CrudianError,
    )
  })
})
